### PR TITLE
Feature/check household match

### PR DESCRIPTION
**Pull request by dwybourn**

### CODE REVIEW

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/EligibilityService.java [hunk 1]
```diff
@@ -43,12 +47,16 @@ public EligibilityResponse checkEligibility(EligibilityRequest eligibilityReques
         String nino = eligibilityRequest.getPerson().getNino();
         Optional<UCHousehold> ucHousehold = ucHouseholdRepository.findHouseholdByAdultWithNino(nino);
         if (ucHousehold.isPresent()) {
-            return createEligibilityResponse(ucHousehold.get());
+            return householdVerifier.detailsMatch(ucHousehold.get(), eligibilityRequest.getPerson())
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/EligibilityService.java]: This and the next ternary feel a bit too big, so I'll refactor them out into a method for clarity.

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import uk.gov.dhsc.htbhf.dwp.entity.legacy.LegacyHousehold;
+import uk.gov.dhsc.htbhf.dwp.entity.uc.UCHousehold;
+import uk.gov.dhsc.htbhf.dwp.model.PersonDTO;
+
+@Service
+public class HouseholdVerifier {
+
+    /**
+     * Determines if a given household contains an adult matching a given person.
+     * DWP rules specify to only match on the first six characters of the address line 1 field.
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java]: This documentation is a bit confusing as you only specify the address match, but in fact it matches on much more, can you detail all the matches it does here please?

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+     * @param household household to check
+     * @param person person to check
+     * @return true the details of the household and the person match.
+     */
+    public Boolean detailsMatch(UCHousehold household, PersonDTO person) {
+        return household.getAdults().stream()
+                .anyMatch(adult ->
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java]: I think this method and the one below need a bit of refactoring, the contents of the `anyMatch` can be refactored out into methods, and also the return statement in the next function for clarity.

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        return nameMatches
+                && firstSixCharacterMatch(household.getAddressLine1(), person.getAddress().getAddressLine1())
+                && areEqual(household.getPostcode(), person.getAddress().getPostcode());
+    }
+
+    private boolean firstSixCharacterMatch(String s1, String s2) {
+        return areEqual(StringUtils.left(s1, 6), StringUtils.left(s2, 6));
+    }
+
+    private Boolean areEqual(String s1, String s2) {
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java:51]: There must be a method in Guava or Apache commons that already does this?

File: api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+import uk.gov.dhsc.htbhf.dwp.model.PersonDTO;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static uk.gov.dhsc.htbhf.dwp.entity.LegacyHouseholdFactory.aHouseholdWithNoAdultsOrChildren;
+
+public class HouseholdVerifierLegacyHouseholdTest {
+
+    private final HouseholdVerifier householdVerifier = new HouseholdVerifier();
+    private final AddressDTO address = AddressDTO.builder()
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java]: I think all the test data objects you create for this test should be in test data factories.

File: api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    void shouldReturnTrueWhenPersonMatchesLegacyHousehold() {
+        LegacyAdult adult = LegacyAdult.builder()
+                .forename("Lisa")
+                .surname("Simpson")
+                .build();
+        LegacyHousehold household = aHouseholdWithNoAdultsOrChildren().build().addAdult(adult);
+
+        Boolean response = householdVerifier.detailsMatch(household, person);
+
+        assertTrue(response);
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java]: I'd use AssertJ for the assertions in this test, they yield much better messages is the tests fail.

File: api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierUCHouseholdTest.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+import uk.gov.dhsc.htbhf.dwp.entity.uc.UCAdult;
+import uk.gov.dhsc.htbhf.dwp.entity.uc.UCHousehold;
+import uk.gov.dhsc.htbhf.dwp.model.AddressDTO;
+import uk.gov.dhsc.htbhf.dwp.model.PersonDTO;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static uk.gov.dhsc.htbhf.dwp.entity.UCHouseholdFactory.aHouseholdWithNoAdultsOrChildren;
+
+public class HouseholdVerifierUCHouseholdTest {
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierUCHouseholdTest.java:11]: Same comments for this test around using test data factories and AssertJ.

**Review by mikeSmith08 — CHANGES_REQUESTED** [review]
See individual comments.

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java (hunk 2 above)
Comment (YetAnotherMatt) [reply api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java]: The description in the javadoc refers specifically to the matching for an address, but we also match on name.

Comment (YetAnotherMatt) [reply api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java]: I think we'll want to be quite explicit on how we match on name & address in the review/showcase. I'm a little worried that first names may not be so easy to match - Matt vs Matthew, etc.

File: api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java (hunk 6 above)
Comment (YetAnotherMatt) [reply api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java]: assertJ also reads better in the test itself

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java (hunk 3 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java]: I'm not sure a method per equality check would read better than what's already there.

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java (hunk 4 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java:51]: None that I could find.

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java (hunk 3 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java]: Done.

File: api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierUCHouseholdTest.java (hunk 7 above)
Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierUCHouseholdTest.java:11]: Done.

File: api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java (hunk 6 above)
Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java]: Done.

File: api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java (hunk 5 above)
Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifierLegacyHouseholdTest.java]: Done.

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java (hunk 2 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/HouseholdVerifier.java]: Updated to include all the checks we perform.

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/EligibilityService.java (hunk 1 above)
Comment (dwybourn) [reply api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/EligibilityService.java]: Done.

**Review by gethyn1 — APPROVED** [review]

**Review by mikeSmith08 — APPROVED** [review]

File: api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/EligibilityService.java [hunk 8]
```diff
@@ -43,15 +47,27 @@ public EligibilityResponse checkEligibility(EligibilityRequest eligibilityReques
...
         }
 
         var response = restTemplate.postForEntity(uri, eligibilityRequest, EligibilityResponse.class);
         return response.getBody();
     }
+
+    private EligibilityResponse getEligibilityResponse(EligibilityRequest eligibilityRequest, UCHousehold ucHousehold) {
+        return householdVerifier.detailsMatch(ucHousehold, eligibilityRequest.getPerson())
+                ? createEligibilityResponse(ucHousehold)
+                : EligibilityResponse.builder().eligibilityStatus(NOMATCH).build();
```
Comment (YetAnotherMatt) [inline api/src/main/java/uk/gov/dhsc/htbhf/dwp/service/EligibilityService.java:65]: might be worth extracting this to a 'createNoMatchResponse'?

**Review by YetAnotherMatt — APPROVED** [review]

**Pull request merged**